### PR TITLE
Remove `chain::Id` completely

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Matching on `chain::Id` is cumbersome because it lacks guarantees about its values. For example, `chain::Id` can be constructed from a `u64` or `U256` without error handling for unsupported chains.

So to match properly, we need to first convert `chain::Id` to `chain::Chain`, and then also handle failure of conversion.

This proposal suggests replacing `chain::Id` entirely with `chain::Chain`, which is already well-encapsulated. This change is feasible because all scenarios where `chain::Chain` cannot be constructed are handled appropriately—using an early `.expect()` whenever a service is restarted.

# Changes

- [ ] Remove `chain::Id` and use `chain::Chain` instead
- [ ] Had to add `Hardhat` chain to make driver tests pass.

## How to test
Existing tests should verify no bugs were introduced.

### CODE REVIEW

File: crates/driver/src/domain/eth/eip712.rs [hunk 1]
```diff
@@ -7,25 +7,3 @@
...
-        let abi_string = ethabi::encode(&[
-            ethabi::Token::Uint(web3::signing::keccak256(fields.type_hash).into()),
-            ethabi::Token::Uint(web3::signing::keccak256(fields.name).into()),
-            ethabi::Token::Uint(web3::signing::keccak256(fields.version).into()),
-            ethabi::Token::Uint(ethereum_types::U256::from(u64::from(fields.chain_id))),
-            ethabi::Token::Address(fields.verifying_contract.into()),
-        ]);
-        Self(web3::signing::keccak256(abi_string.as_slice()))
-    }
-}
```
Comment (squadgazzz) [inline crates/driver/src/domain/eth/eip712.rs:31]: Just to double-check, this is not related to this PR and just removes previously unused code, right?

**Comment by squadgazzz** [comment]
> Had to add Hardhat chain to make driver tests pass.

I didn't quite get from the description and the code where the hardhat chain id comes from.

File: crates/driver/src/infra/blockchain/mod.rs [hunk 2]
```diff
@@ -290,6 +294,7 @@ impl Error {
             }
             Error::GasPrice(_) => false,
             Error::AccessList(_) => true,
+            Error::UnsupportedChain => false,
```
Comment (MartinquaXD) [inline crates/driver/src/infra/blockchain/mod.rs:297]: A new error variant gets added which can never appear here and yet we have to handle this case. This is a code smell and indicates that these functions should return different errors.
Something for a future refactor, though.

File: crates/shared/src/sources/mod.rs [hunk 3]
```diff
@@ -68,6 +68,7 @@ pub fn defaults_for_network(chain: &Chain) -> Vec<BaselineSource> {
             BaselineSource::UniswapV3,
         ],
         Chain::Sepolia => vec![BaselineSource::TestnetUniswapV2],
+        Chain::Hardhat => vec![BaselineSource::TestnetUniswapV2],
```
Comment (MartinquaXD) [inline crates/shared/src/sources/mod.rs:71]: This should be empty, no? Test nodes don't deploy any liquidity on their own so there is no default liquidity we could use here.

File: crates/chain/src/lib.rs [hunk 4]
```diff
@@ -59,6 +64,8 @@ impl Chain {
             Self::Sepolia => Duration::from_millis(12_000),
             Self::ArbitrumOne => Duration::from_millis(250),
             Self::Base => Duration::from_millis(2_000),
+            // dummy value as hardhat chain doesn't have a fixed block time
+            Self::Hardhat => Duration::from_millis(1_337),
```
Comment (MartinquaXD) [inline crates/chain/src/lib.rs:68]: Do the tests pass if you put `panic!("no default values for <PARAMETER> available on hardhat")` for all the cases you currently return dummy values for hardhat?
I think hitting the default code paths would be an error, no?

File: crates/driver/src/domain/eth/eip712.rs (hunk 1 above)
Comment (sunce86) [reply crates/driver/src/domain/eth/eip712.rs:31]: True. I noticed it because the change touches `chain_id`

**Comment by sunce86** [comment]
> > Had to add Hardhat chain to make driver tests pass.
> 
> I didn't quite get from the description and the code where the hardhat chain id comes from.

https://hardhat.org/hardhat-network/docs/reference#chainid

File: crates/shared/src/sources/mod.rs (hunk 3 above)
Comment (sunce86) [reply crates/shared/src/sources/mod.rs]: Adjusted.

File: crates/chain/src/lib.rs (hunk 4 above)
Comment (sunce86) [reply crates/chain/src/lib.rs]: Adjusted. Should be fine that way as well. Let's see if tests pass

File: crates/driver/src/infra/config/file/load.rs [hunk 5]
```diff
@@ -40,9 +41,12 @@ pub async fn load(chain: chain::Id, path: &Path) -> infra::Config {
     });
 
     assert_eq!(
-        config.chain_id.map(chain::Id::from).unwrap_or(chain),
+        config
+            .chain_id
+            .and_then(|id| Chain::try_from(id).ok())
```
Comment (mstrug) [inline crates/driver/src/infra/config/file/load.rs:46]: Using `ok()` here discards `ChainIdNotSupported` error so message log would be misleading that chain id doesn't match connected node instead of not supported chain id.

File: crates/shared/src/bad_token/token_owner_finder/blockscout.rs [hunk 6]
```diff
@@ -26,6 +26,7 @@ impl BlockscoutTokenOwnerFinder {
             Chain::Sepolia => "https://eth-sepolia.blockscout.com/api",
             Chain::ArbitrumOne => "https://arbitrum.blockscout.com/api",
             Chain::Base => "https://base.blockscout.com/api",
+            Chain::Hardhat => return Err(anyhow::anyhow!("Hardhat chain not supported")),
```
Comment (mstrug) [inline crates/shared/src/bad_token/token_owner_finder/blockscout.rs:29]: As we are using anyhow, let's use dedicated macro for early return:
`anyhow::bail!("Hardhat chain not supported")`.

**Comment by squadgazzz** [comment]
> [hardhat.org/hardhat-network/docs/reference#chainid](https://hardhat.org/hardhat-network/docs/reference#chainid)

I mean, why do we need It in the code now. Some of the tests use this chain id implicitly?

File: crates/driver/src/infra/blockchain/mod.rs (hunk 2 above)
Comment (sunce86) [reply crates/driver/src/infra/blockchain/mod.rs:297]: I don't have a strong opinion here. If you look at how the same constructor is used in `autopilot`, you can see that `.expect()` is used when the `Chain::try_from` is called in `run.rs` (so when we are aware of our surrounding and we explicitly want to exit), and error type is used inside `Rpc::new` because we don't know when and how this function will be called.

But I also agree that this error is rarely hit. If we agree to panic everywhere, then we can remove `ChainIdNotSupported` error from `chain` crate altogether. This is dangerous since we might want to try to build the `Chain` object from unreliable source that could give garbage, but it's not the case now.

File: crates/chain/src/lib.rs [hunk 7]
```diff
@@ -59,6 +64,7 @@ impl Chain {
             Self::Sepolia => Duration::from_millis(12_000),
             Self::ArbitrumOne => Duration::from_millis(250),
             Self::Base => Duration::from_millis(2_000),
+            Self::Hardhat => panic!("unsupported block time for Hardhat chain"),
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/chain/src/lib.rs:67]: Yup, it's ok since no settlement:new is called in the driver.

**Comment by sunce86** [comment]
> return Err(anyhow::anyhow!("Hardhat chain not supported")),

Driver tests do. ☝️

File: crates/driver/src/infra/config/file/load.rs (hunk 5 above)
Comment (sunce86) [reply crates/driver/src/infra/config/file/load.rs]: You are right! Adjusted.

File: crates/driver/src/infra/blockchain/mod.rs (hunk 2 above)
Comment (mstrug) [reply crates/driver/src/infra/blockchain/mod.rs:297]: So another option would be to introduce new error type dedicated for `Rpc` object, for instance `RpcError`, which will contain only one variant `ChainIdNotSupported`, and caller of the constructor will decide if want to panic or map error to other type.
Alternatively constructor can return boolean type error as there is only one function which can fail with one error code.

File: crates/driver/src/infra/blockchain/mod.rs [hunk 8]
```diff
@@ -27,7 +28,8 @@ impl Rpc {
     /// at the specifed URL.
     pub async fn new(url: &url::Url) -> Result<Self, Error> {
```
Comment (mstrug) [inline crates/driver/src/infra/blockchain/mod.rs:29]: Please rename the constructor to `try_new()` to follow Rust coding convention.

Comment (sunce86) [reply crates/driver/src/infra/blockchain/mod.rs:29]: There is a lot of similar constructors that are returning `Result`. I think we can do this in a separate PR, it could be a good first issue.

https://github.com/cowprotocol/services/issues/3179

File: crates/driver/src/infra/blockchain/mod.rs (hunk 2 above)
Comment (squadgazzz) [reply crates/driver/src/infra/blockchain/mod.rs:297]: > So another option would be to introduce new error type dedicated for Rpc object, for instance RpcError, which will contain only one variant ChainIdNotSupported, and caller of the constructor will decide if want to panic or map error to other type.

That seems reasonable. @sunce86 do you mind implementing it?

Comment (sunce86) [reply crates/driver/src/infra/blockchain/mod.rs]: Sure, I created a new error type

**Review by squadgazzz — APPROVED** [review]
LGTM
